src/components/syncplay/syncplayManager.js

```diff
@@ -0,0 +1,737 @@
+                break;
+        }
+    }
+
+    /**
+     * Prepares this client to join a group by loading the required content.
+     * @param {Object} apiClient The ApiClient.
+     * @param {string} groupId The group to join.
+     * @param {Object} sessionData Info about the content to load.
+     */
+    prepareSession (apiClient, groupId, sessionData) {
+        var serverId = apiClient.serverInfo().Id;
+        playbackManager.play({
+            ids: sessionData.ItemIds,
+            startPositionTicks: sessionData.StartPositionTicks,
+            mediaSourceId: sessionData.MediaSourceId,
+            audioStreamIndex: sessionData.AudioStreamIndex,
+            subtitleStreamIndex: sessionData.SubtitleStreamIndex,
+            startIndex: sessionData.StartIndex,
+            serverId: serverId
+        }).then(() => {
+            waitForEvent(this, "PlayerChange").then(() => {
+                playbackManager.pause();
+                var sessionId = getActivePlayerId();
+                if (!sessionId) {
+                    console.error("Missing sessionId!");
+                    toast({
+                        // TODO: translate
+                        text: "Failed to enable Syncplay! Missing session id."
+                    });
+                    return;
+                }
+                // Get playing item id
+                let playingItemId;
+                try {
+                    const playState = playbackManager.getPlayerState();
+                    playingItemId = playState.NowPlayingItem.Id;
+                } catch (error) {
+                    playingItemId = "";
+                }
+                // Sometimes JoinGroup fails, maybe because server hasn't been updated yet
+                setTimeout(() => {
+                    apiClient.sendSyncplayCommand(sessionId, "JoinGroup", {
+                        GroupId: groupId,
+                        PlayingItemId: playingItemId
+                    });
+                }, 500);
+            });
+        }).catch((error) => {
+            console.error(error);
+            toast({
+                // TODO: translate
+                text: "Failed to enable Syncplay! Media error."
+            });
+        });
+    }
+
+    /**
+     * Enables Syncplay.
+     * @param {Object} apiClient The ApiClient.
+     * @param {Date} enabledAt When Syncplay has been enabled. Server side date.
+     * @param {boolean} showMessage Display message.
+     */
+    enableSyncplay (apiClient, enabledAt, showMessage = false) {
+        this.syncplayEnabledAt = enabledAt;
+        this.injectPlaybackManager();
+        events.trigger(this, "SyncplayEnabled", [true]);
+
+        waitForEvent(this, "SyncplayReady").then(() => {
+            this.processCommand(this.queuedCommand, apiClient);
+            this.queuedCommand = null;
+        });
+        this.syncplayReady = false;
+        this.notifySyncplayReady = true;
```

Comment:
I feel like it's not that relevant as the event is called asynchronously (this function has to terminate before the event can be triggered). But even if it wasn't, it makes more sense to first register the event listener and then tell the trigger code that the event can be triggered, because the listener has been set.

That's my view on it :sweat_smile:. The code can be rearranged but the flow is clearer this way IMO.